PIP-9:-Topic-Compaction.md

```diff
@@ -0,0 +1,69 @@
+* **Status**: Proposal
+* **Authors**: Ivan Kelly
+* **Pull Request**: [ ]
+* **Mailing List discussion**:
+* **Tasks break down**:
+* **Prototype**: https://github.com/ivankelly/incubator-pulsar/tree/compaction-prototype
+
+## Motivation
+
+In some usecases, a pulsar client will want to read the whole history of a log to get an up to date image. If the log is very long this can take a long time. The client will also end up reading a lot of useless data, as some entries will be obscured by later entries.
+
+To avoid this problem we need to do log compaction. Periodically some process will go through the backlog of a topic and remove messages which would be obscured by a later message, i.e. there is a later message in the topic with the same key.
+
+## Design
+
+This document is concerned with the compaction mechanism, which takes a topic, compacts it down, and notifies the broker that compaction has occurred. It is no concerned with the scheduling of compaction. This can be added in a later proposal. Compaction only applies to persistant topics.
+
+To trigger compaction on a topic, the user calls
+```
+$ bin/pulsar-admin persistent compact <TOPIC_URI>
+```
+This command will kick off a compaction process which runs on the machine that the command is run from.
+
+The Pulsar compaction mechanism is very similar to that of Kafka. The compactor will first iterate over the message of the topic from the start. For each key it finds it will keep a record of the latest occurance of that key. Once it gets to the end of the topic, it will create a BookKeeper ledger to store the compacted topic. It will iterate over the topic a second time. During the second iteration, it will check the key of each message. If the key matches the latest instance of that key or the message has no key, the message Id, metadata and payload is written to the compacted topic ledger. Once the second run reaches the end point of the first run, compaction terminates. The compacted topic ledger is sealed/closed, and the ID of the ledger is written to the metadata of the topic, along with the message ID of the last message compacted (we'll call this the compaction horizon).
+
+If compaction is enabled for the topic, the pulsar broker serving the topic will watch the metadata for changes to the compacted topic ledger ID. When a client requests to read a message, the ID of the message to read is compared to the compaction horizon. If the ID is higher than or equal to the horizon, nothing new happens. If the ID is lower than the compaction horizon, then the read is served from the compaction horizon. As the compacted topic ledger does not contain all messages, the specified ID may not exist. When reading from the compacted topic ledger, we do a binary search to find the first message with a ID greater or equal to the ID specified in the read request.
```

Comment:
>>find the first message with a ID greater or equal to the ID specified in the read request.
It would be good to be explicit about why this would be the right action

Reply:
Will update.